Examples/Nodify.Shapes/Canvas/CanvasViewModel.cs

```diff
@@ -0,0 +1,249 @@
+        public IActionsHistory UndoRedo { get; } = ActionsHistory.Global;
+
+        private readonly Random _rand = new Random();
+
+        public CanvasViewModel()
+        {
+            CanvasToolbar = new CanvasToolbarViewModel(this);
+
+            UndoCommand = new RequeryCommand(UndoRedo.Undo, () => UndoRedo.CanUndo);
+            RedoCommand = new RequeryCommand(UndoRedo.Redo, () => UndoRedo.CanRedo);
+
+            MoveShapesStartedCommand = new DelegateCommand(MoveShapesStartedHandler);
+            MoveShapesCompletedCommand = new DelegateCommand(MoveShapesCompletedHandler);
+
+            SelectShapesStartedCommand = new DelegateCommand(SelectShapesStartedHandler);
+            SelectShapesCompletedCommand = new DelegateCommand(SelectShapesCompletedHandler);
+
+            ResizeShapeStartedCommand = new DelegateCommand(ResizeShapeStartedHandler);
+            ResizeShapeCompletedCommand = new DelegateCommand(ResizeShapeCompletedHandler);
+
+            DeleteSelectionCommand = new DelegateCommand(DeleteSelection, () => !CanvasToolbar.Locked);
+
+            CreateConnectionCommand = new DelegateCommand<(object Source, object Target)>(
+                ((object Source, object Target) pendingConnection) => AddConnection((ConnectorViewModel)pendingConnection.Source, (ConnectorViewModel)pendingConnection.Target),
+                ((object Source, object Target) pendingConnection) => CanConnect((ConnectorViewModel)pendingConnection.Source, (ConnectorViewModel)pendingConnection.Target));
+
+            RemoveConnectionCommand = new DelegateCommand<ConnectionViewModel>(RemoveConnection);
+
+            SelectedShapes.WhenAdded(shape =>
+            {
+                ShapeToolbar.Shape = SelectedShapes.Count == 1 ? shape : null;
+            });
+
+            SelectedShapes.WhenRemoved(shape =>
+            {
+                ShapeToolbar.Shape = SelectedShapes.Count == 1 ? SelectedShapes.Single() : null;
+            });
+
+            SelectedShapes.WhenCleared(shapes => ShapeToolbar.Shape = null);
+
+            FillCanvasWithShapes();
+        }
+
+        private void MoveShapesStartedHandler()
+        {
+            UndoRedo.ExecuteAction(new MoveShapesAction(this));
+            ShapeToolbar.Hide();
+        }
+
+        private void MoveShapesCompletedHandler()
+        {
+            if (UndoRedo.Current is MoveShapesAction movesShapes)
+            {
+                movesShapes.SaveLocations();
+            }
+
+            ShapeToolbar.Show();
+        }
+
+        // TODO: This does not fire for single selection (ItemContainer click), therefore I added IsSelected to ShapeViewModel.
+        private void SelectShapesStartedHandler()
+        {
+            // The workaround is to record all IsSelected changes into one history item using History.Pause and History.Resume
+            UndoRedo.Pause("Select shapes");
```

Comment:
The text label on batch/pause is another debatable thing, where i think something more structured would be nicer (like the overall action itself or enum(likely too constricting)). However i think its likely fine as that way its easily usable.

Reply:
The label is meant to be displayed in the UI similar to how Unreal Engine is doing it:
![image](https://github.com/miroiu/nodify/assets/12727904/dd44ef75-5d8c-49b1-b61d-45e8bc161d96)